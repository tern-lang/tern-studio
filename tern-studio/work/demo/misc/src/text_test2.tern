
let t = replaceText("let x = `hello world`;");

println(t);


func replaceText(text: String) {
   let index = 0;
   let findFirst = (text, start) -> {
      let min = -1;
      
      for(open in ["'", "`", "\""]) {
         let pos = text.indexOf(open, start);
         
         if(pos > 0 && text[pos-1] == '\\') {
            continue;
         }
         if(pos != -1) {
            if(min == -1) {
               min = pos;
            } else  if(pos < min) {
               min = pos;
            }
         }
      }
      return min;
   };
   
   while(index != -1) {
      index = findFirst(text, index);
      
      if(index == -1 || index >= text.length) {
         break;
      }
      const quote = text[index];
      const end = text.indexOf(quote, index + 1);
      
      if(end == -1) {
         break;
      }
      const prefix = text.substring(0, index);
      const suffix = text.substring(end + 1);
      const token = text.substring(index, end + 1);
      const encoded = encode(token);
      const replace = "<span style='color: #8959A8'>${encoded}</span>";
      
      text = "${prefix}${replace}${suffix}";
      index += replace.length;
   }
   return text;
}

func encode(text: String) {
   let builder = StringBuilder();
   
   for(i in 0 to text.length - 1){
      let ch = text[i];
      
      if(Character.isLetter(ch)) {
         let code = ch as Integer;
         builder.append("&#${code}");
      } else {
         builder.append(ch);
      }
   }
   return builder.toString();
}
