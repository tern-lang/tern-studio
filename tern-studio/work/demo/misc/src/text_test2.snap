
func foo(){
   println("x");
}
foo();
